retriever/lib/engine.py

```diff
@@ -438,70 +448,56 @@ def download_file(self, url, filename):
-                # if no local copy, download the data
-                self.create_raw_data_dir()
-                if not downloaded:
-                    self.download_file(url, archivename)
-                    downloaded = True
-
-                if filetype == 'zip':
-                    try:
-                        archive = zipfile.ZipFile(archivename)
-                        if archive.testzip():
-                            # This fixes an issue with the zip files that was causing errors on
-                            # Python 3. testzip() returns the names of any files with issues so if
-                            # it exists there is a problem. For details of the issue and the fix see:
-                            # see """https://stackoverflow.com/questions/41492984/
-                            # zipfile-testzip-returning-different-results-on-python-2-and-python-3"""
-                            archive.getinfo(filename).file_size += (2 ** 64) - 1
-                        open_archive_file = archive.open(filename, 'r')
-                    except zipfile.BadZipFile as e:
-                        print("\n{0} can't be extracted, may be corrupt \n{1}".format(filename, e))
-
-                elif filetype == 'gz':
-                    # gzip archives can only contain a single file
-                    open_archive_file = gzip.open(archivename, 'r')
-                elif filetype == 'tar':
-                    archive = tarfile.open(filename, 'r')
-                    open_archive_file = archive.extractfile(filename)
-
-                fileloc = self.format_filename(os.path.join(archivebase, filename))
-                fileloc = os.path.normpath(fileloc)
-                if not os.path.exists(os.path.dirname(fileloc)):
-                    os.makedirs(os.path.dirname(fileloc))
-
-                unzipped_file = open(fileloc, 'wb')
-                for line in open_archive_file:
-                    unzipped_file.write(line)
-                open_archive_file.close()
-                unzipped_file.close()
-                if 'archive' in locals():
-                    archive.close()
-
-    def drop_statement(self, objecttype, objectname):
+            archive_base = os.path.splitext(os.path.basename(archive_name))[0]
+            archive_dir = os.path.join(DATA_WRITE_PATH, archive_base)
+            archive_dir = archive_dir.format(dataset=self.script.name)
+
+        if not file_names:
+            self.download_file(url, archive_name)
+            if archive_type == 'tar' or archive_type == 'tar.gz':
+                file_names = self.extract_tar(archive_full_path, archive_dir, archive_type)
+            elif archive_type == 'zip':
+                file_names = self.extract_zip(archive_full_path, archive_dir)
+            elif archive_type == 'gz':
+                file_names = self.extract_gz(archive_full_path, archive_dir)
+            return file_names
+
+        if file_names:
```

Comment:
This is a little confusing since we define file_names in the previous `if` and if you miss the `return` it looks like this code should always run. I might use an `else` where instead.